text/2022-05-12-multi-cluster-support.md

```diff
@@ -0,0 +1,328 @@
+      -> Apply and return
+   -> If the finalizer doesn't exist, add a finalizer
+   -> Get the chaos in child cluster
+      -> If the chaos in the child cluster doesn't exist, create the chaos in the child cluster
+      -> If the chaos in the child cluster exists, read the `Status` of the chaos in the child cluster
+         and update the `Status` of the chaos in the parent cluster
+```
+
+### Q&A
+
+1. Why should we install the chaos mesh controller in every cluster? Wouldn’t it
+   be better to manage all chaos in a single cluster (even process)?
+
+   It sounds great to listen and respond to every chaos in a single process,
+   which means we will only need to install chaos daemon in the target cluster.
+   However, the reality is that there is no common way to communicate with the
+   chaos daemon in another cluster. I have thought about proxy, port-forward or
+   other things, but nothing is reliable enough.
+
+   I also don’t want to depend on the user to tell us the network topology: the
+   clusters are in the same network / different network / proxy / ingress / … .
+   It’s too hard for both of us and users. If we operate the chaos through
+   another controller in the target cluster, we only need an available
+   kubeconfig, which is much simpler than the network configuration.
+
+2. Why is the `remoteCluster` an extra field? Would a standalone resource /
+   annotation / … be better?
+
+   I can’t say which one is better. It’s a blind choice. I don’t want to create
+   a new embedded standalone resource because the existing two embedded
+   resources: Schedule and Workflow are too huge (but it’s actually not a strong
+   evidence). I prefer fields over annotations because fields sound more like an
+   official API, but annotations do not (or maybe we could start from
+   annotations? I’m not sure).
+
+3. Can one chaos inject fault into different clusters?
+
+   Obviously, no. I don’t think we can provide a general way to implement this
+   in the near future.
+
+4. Where will the webhook run?
+
+   In the parent chaos mesh. Because we will not create chaos in the webhook
+   (but in the reconcile), the validation should be done in the parent’s
+   webhook.
+
+5. What about chaos events?
+
+   Thanks to @xlgao-zju , we have stored some events inside the chaos status, so
+   we don’t need to synchronize the event between two clusters.
+
+## Authentication / Authorization
+
+The Chaos Mesh project has paid a lot of attention to Authentication and
+Authorization. In most situations, we choose to follow the RBAC pattern. The
+multi-cluster injection will also follow the RBAC pattern. Like what we have
+done inside the validate_auth.go, if the remoteCluster is not nil, a webhook
+will create a SubjectAccessReview to check whether the user is able to create
+this object in the target cluster/namespace.
+
+The only problem is that the username could be different among the clusters. To
```

Comment:
> The only problem is that the username could be different among the clusters.

Not only the username but also the user credential.

It requires users to provide different credentials for different clusters.  It seems we would save the credentials into `userMap`.

When the "Manager" sync the *Chaos from "ParentCluster" to "RemoteCluster", it requires a certain kubeconfig with users' credential. (Or maybe we just use the given kubeconfig 🤷‍♂️)

Reply:
> Not only the username but also the user credential.

I think chaos mesh itself could use a super powerful credential, and impersonate any one (with only `username`) to create things.

Or is it possible to only validate the authorization only in the parent cluster? As the chaos controller in the parent cluster could also create `SubjectAccessReview` in the remote cluster. I think it's better as the authorization error will block the user creating the resource (which he don't actually have privilege to create).